docs/proposals/20230525-openstack-microversion-support.md

```diff
@@ -0,0 +1,133 @@
+# Configurable/multi microversion support
+
+OpenStack is diverse.
+Multiple versions and configurations exists that can impact what microversions and features that are supported.
+The user (of CAPO) can also to some extent choose what features to make use of.
+The cross section of this is where CAPO is useful:
+
+## Motivation
+
+A CAPO cluster can be configured in many ways and depending on this configuration, certain features may or may not be needed.
+This makes it necessary to know and detect if a feature is used that requires a certain microversion.
+
+It is not enough to check if the user has asked for a specific configuration that would require a certain microversion.
+The requirement can come from the OpenStack instance itself.
+As an example, we have seen environments where all volumes are multi-attachable.
+I.e. it is not possible to create a volume that is not multi-attachable, and thus multi-attach support is a must to be able to do anything in this environment, even if the user doesn't explicitly ask for it.
+
+Currently CAPO has a fixed microversion set.
+It is a take it or leave it situation.
+Either this version is supported by the OpenStack environment or you cannot use CAPO.
+
+This proposal is to make CAPO more flexible in what it supports.
+For example, there could be other versions that doesn't really impact CAPO in any way or there could be cross sections between what the OpenStack environment and CAPO supports, that could be useful to the user if certain features are avoided.
+Here is a simple diagram to try to summarize what this proposal aims to achieve:
+
+```text
+  ┌────────────────────────────────────┐
+  │                                    │
+  │                                    │
+  │    Versions supported by           │
+  │    OpenStack instance              │
+  │                                    │
+  │    ┌─────────────────────────┐     │
+  │    │                         │     │
+  │    │  Versions available to  │     │
+  │    │  the user               │     │
+  └────┼─────────────────────────┼─────┘
+       │                         │
+       │  Versions supported     │
+       │  by CAPO                │
+       │                         │
+       └─────────────────────────┘
+```
+
+"Versions" could be changed to "features" in the diagram also, but a version difference does not always expose a difference in features at the CAPO level.
+Therefore "versions" is used in the diagram to highlight the need to support multiple versions even though that may not give any new features.
+
+### Goals
+
+1. Detect user requested features and their version needs.
+
+2. Detect server version support.
+
+### Non-goals
+
+- To allow the user to set the versions freely.
+
+### CAPO microversions history
+
+Initially no microversions were specified (only versions).
+This is still true for all clients except compute.
```

Comment:
I think only Cinder, Manila and Barbicna support microversions beside Nova. Neutron uses extensions, Keystone does neither. It's probably worth mentioning that here.

Reply:
Thanks! Added this. Could you double check if it makes sense?